### ISSUE TITLE
HAProxy - CA/CRL chain becomes broken in time under heavy load

### ISSUE THREAD

**Issue by pkubicsek** [haproxy/haproxy#38]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 1.8.4-1deb90d 2018/02/08
Copyright 2000-2018 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -m64 -march=x86-64 -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_SLZ=1 USE_REGPARM=1 USE_OPENSSL=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace

Linux *redacted* 3.10.0-327.18.2.el7.x86_64 #1 SMP Thu May 12 11:03:55 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

```

## What's the configuration?

```
#### global
   chroot                              /var/lib/haproxy
   stats socket                        *redacted*
   user                                haproxy
   group                               haproxy
   daemon
   debug
   log                                 *redacted* local2
   maxconn                             50000
   tune.ssl.default-dh-param           2048
   log-send-hostname
   ssl-default-bind-options no-sslv3
   ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS
   ssl-default-server-options no-sslv3
   ssl-default-server-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS

#### defaults
   mode                                http
   log                                 global
   option                              httplog
   option                              dontlognull
   option                              forwardfor  except 127.0.0.0/8
   option                              redispatch
   retries                             3
   timeout connect                     5s
   timeout client                      60s
   timeout server                      60s
   timeout queue                       5s
   timeout http-request                30s
   timeout http-keep-alive             15s
   timeout check                       10s
   timeout tunnel                      120m
   timeout client-fin                  5s
   timeout server-fin                  5s

#### frontend  frontend_ssl_auth
   bind *redacted* ssl crt /etc/ssl/hostname_le.pem no-sslv3 ca-file /etc/ssl/clients_chain.pem verify required crl-file /etc/ssl/doxxbet/clients_chain.pem
   log-format %ci:%cp\ [%t]\ {%{+Q}[ssl_c_s_dn(cn)],%[ssl_c_verify]}\ %ft\ %b/%s\ %Tq/%Tw/%Tc/%Tr/%Tt\ %ST\ %B\ %CC\ %CS\ %tsc\ %ac/%fc/%bc/%sc/%rc\ %sq/%bq\ %hr\ %hs\ %{+Q}r
   maxconn 10000
   default_backend be_backend1

#### backend be_backend1
  balance roundrobin
   fullconn 4000
   default-server maxconn 1000 inter 5s rise 2 fall 4
   option httpchk GET /healthcheck
   http-check expect status 200
   server server_1 *redacted* check weight 10
   server server_2 *redacted* check weight 10
   server server_3 *redacted* check weight 10
   server server_4 *redacted* check weight 10
   http-request set-header X-Forwarded-Port %[dst_port]
   http-response del-header Server
   http-response del-header X-AspNet-Version
   http-response del-header X-Powered-By
   http-response del-header X-Application-Context
```

## Steps to reproduce the behavior

1. There are 27 CAs with 27 corresponding CRLs and everything works as expected when we start haproxy, because we are using complex CA structure with multiple sub_ca trees and so on. 
2. Server is processing a lot of requests (around 20 mil per day), some of them are long-polling SignalR connections and some of them are normal http/https. There are also around 20 frontends and around 50 backends. Some frontends using 2-way ssl some not.
3. When clients chains is changed (for example reissued CRL) we update it on filesystem and restart haproxy. This mechanism is done via script in crontab on regular basis, but i've already checked logs and restarts doesn't correspond with our issue
4. Do a load of requests and just wait :( It's hard to reproduce, because it doesn't have connection to anything special, at least I didn't realise it.

This issue was seen also in older releases of HAProxy, starting from  version 

## Actual behavior

When HAProxy started, all works as expected. Sometimes it's working around few weeks, sometimes more but sometimes less. Then something happens and frontend start to decline clients as not trusted 

```
Feb  8 10:20:10 redacted haproxy[20999]: redacted:22941 [08/Feb/2019:10:20:10.345] frontend_ssl_auth/1: SSL client certificate not trusted
```

Funny thing is, it affects only just some sub_ca's in clients_chain. Some clients are working and some not, depends on which CA signed theirs certificates. 

We already checked clients_chain.pem file when this situation happens and file is OK. It seems to me like chain data in memory is broken. Fix is simple, just restart whole haproxy, but we would like to have this issue fixed because it's painfull for us. 

## Do you have any idea what may have caused this?

It seems to me some part of HAProxy in some situation write something to memory part where certificates were stored, so it brakes some CAs in chain, or maybe CRLs.

**Comment by lukastribus** [comment]
> HA-Proxy version 1.8.4-1deb90d 2018/02/08

Gonna need you to upgrade to latest stable. [Haproxy 1.8.4 contains 232 bugs](http://www.haproxy.org/bugs/bugs-1.8.4.html) that are already fixed, so upgrading to latest stable has to be the first step here.


> This issue was seen also in older releases of HAProxy, starting from version

This sentence is incomplete. Since what version you are seeing this behavior and more precisely: what is the youngest released that worked fine for you?

**Comment by pkubicsek** [comment]
Sorry, 

I can do upgrade to the latest version, but I've already check change list and it doesn't seems to were fixed there. But I'll do that. It will take some time, because it takes few weeks when bug shows off.

This issue was also seen in older releases of HAProxy (1.5.18, 1.7.2) which we were using before.

**Comment by wtarreau** [comment]
On Sun, Feb 10, 2019 at 09:42:46AM -0800, pkubicsek wrote:
> I can do upgrade to the latest version,

The only reason for delibereately not applying bugfixes is when you're
certain you have zero problem. You claim the issue is not fixed. Fine.
I claim it is. Who's right ? Shall we close this issue ?

**Comment by lukastribus** [comment]
@wtarreau I think you missunderstood, he is actually upgrading.

@pkubicsek so what you are saying is that you have seen this issue in all releases - you are not aware of any releases that did not have this behavior, right?

**Comment by pkubicsek** [comment]
@lukastribus yes, I will upgrade ... but as you said, I've seen the same issue at least on 3-4 versions we used before. Yes, I dunno any release that didn't have this issue.  I'll be back in few days.

**Comment by wtarreau** [comment]
On Sun, Feb 10, 2019 at 03:21:18PM -0800, Lukas Tribus wrote:
> @wtarreau I think you missunderstood, he is actually upgrading.

Oops, my apologizes. I parsed "I can do upgrade" as "I can not upgrade".
Great then at least we know that if a fix has to be worked on, it will not be for nothing :-)

**Comment by pkubicsek** [comment]
Hello guys,

Just for an update ... update done to HA-Proxy version 1.8.19 2019/02/11. Now I need to wait if it happens or not. Thanks for patience.

**Comment by pkubicsek** [comment]
Hi,

So it happened again today on latest version. Same behaviour like I described before.

```
HA-Proxy version 1.8.19 2019/02/11
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -m64 -march=x86-64 -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_SLZ=1 USE_REGPARM=1 USE_OPENSSL=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
Running on OpenSSL version : OpenSSL 1.0.2k-fips  26 Jan 2017
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : SSLv3 TLSv1.0 TLSv1.1 TLSv1.2
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
```

**Comment by wtarreau** [comment]
Many thanks for verifying! At least now we can rule out all the old bugs. Do you suspect it always happens only after a certain amount of time for example ? If so, one thing that could be interesting to try would be to comment out the "chroot" statement, in case openssl suddenly decides to try to access something on the file system (openssl.cnf, /dev/urandom, etc). I must say I've never seen this problem happen so it could come from anything...

**Comment by pkubicsek** [comment]
Hi,

Yes, it always happens after some time, but this period is non-deterministic. Sometimes it was working few weeks without any issue and sometimes it broke in few days. 

Also important fact is that it only happening to frontends which have two-way ssl enabled. Debug is also very hard because i've never seen this problem on small setup, only in production where we have a lot of requests coming in.

I'll try to disable "chroot" statement on all instances of haproxy we've and let you know if it helped. 

Do you have any more idea what else I can do, or check. I have some strace/tcpdump but I don't see any relevant information there, just that the client was not trusted to haproxy. But I don't know how to check what chain haproxy had loaded in that moment. I can check just pem file on disk, which is OK.

**Comment by pkubicsek** [comment]
Hello,

Nothing changed. It's still happening.

**Comment by kurko** [comment]
Wondering if this is still happening. @pkubicsek how did you proceed with this issue?

**Comment by capflam** [comment]
I'm closing this issue. Feel free to reopen it if necessary. Thanks !

**Comment by stefaro** [comment]
I know this is a really old issue, but I'll add that this could also be caused by a problem with your CRL updates. Since you have many CA and CRL checking activated, it may be that one of them is expired for whatever reason. In my opinion this would cause exactly that issue you are describing. 🤔

**Comment by git001** [comment]
@stefaro I would suggest do open a new issue and add this issue as reference if it makes sense.
